packages/pros/src/devices/peripherals.rs

```diff
@@ -0,0 +1,146 @@
+    pub smart_21: SmartPort,
+
+    pub adi_a: AdiPort,
+    pub adi_b: AdiPort,
+    pub adi_c: AdiPort,
+    pub adi_d: AdiPort,
+    pub adi_e: AdiPort,
+    pub adi_f: AdiPort,
+    pub adi_g: AdiPort,
+    pub adi_h: AdiPort,
+}
+
+impl Peripherals {
+    const unsafe fn new() -> Self {
+        Self {
+            smart_1: SmartPort::new(1),
+            smart_2: SmartPort::new(2),
+            smart_3: SmartPort::new(3),
+            smart_4: SmartPort::new(4),
+            smart_5: SmartPort::new(5),
+            smart_6: SmartPort::new(6),
+            smart_7: SmartPort::new(7),
+            smart_8: SmartPort::new(8),
+            smart_9: SmartPort::new(9),
+            smart_10: SmartPort::new(10),
+            smart_11: SmartPort::new(11),
+            smart_12: SmartPort::new(12),
+            smart_13: SmartPort::new(13),
+            smart_14: SmartPort::new(14),
+            smart_15: SmartPort::new(15),
+            smart_16: SmartPort::new(16),
+            smart_17: SmartPort::new(17),
+            smart_18: SmartPort::new(18),
+            smart_19: SmartPort::new(19),
+            smart_20: SmartPort::new(20),
+            smart_21: SmartPort::new(21),
+
+            adi_a: AdiPort::new(1, None),
+            adi_b: AdiPort::new(2, None),
+            adi_c: AdiPort::new(3, None),
+            adi_d: AdiPort::new(4, None),
+            adi_e: AdiPort::new(5, None),
+            adi_f: AdiPort::new(6, None),
+            adi_g: AdiPort::new(7, None),
+            adi_h: AdiPort::new(8, None),
+        }
+    }
+
+    pub fn take() -> Option<Self> {
+        if PERIPHERALS_TAKEN.swap(true, core::sync::atomic::Ordering::AcqRel) {
+            None
+        } else {
+            Some(unsafe { Self::new() })
+        }
+    }
+
+    pub unsafe fn steal() -> Self {
+        PERIPHERALS_TAKEN.store(true, core::sync::atomic::Ordering::Release);
+        Self::new()
+    }
+}
+
+/// Guarentees that ports are only used once **at runtime**
+/// This is useful for when you want to store a peripherals struct for use in multiple functions.
+/// When possible, use [`Peripherals`] instead.
+pub struct DynamicPeripherals {
+    smart_ports: [bool; 21],
+    adi_slots: [bool; 8],
+}
+impl DynamicPeripherals {
+    /// Creates a new dynamic peripherals
+    /// In order to guarentee that no ports created by this struct,
+    /// this function takes a [`Peripherals`].
+    /// This guarentees safety because [`Peripherals`] cannot be passed by value
+    /// after they have been used to create devices.
+    pub const fn new(_peripherals: Peripherals) -> Self {
+        let smart_ports = [false; 21];
+        let adi_slots = [false; 8];
+        Self {
+            smart_ports,
+            adi_slots,
+        }
+    }
+
+    /// Creates a [`SmartPort`] only if one has not been created on the given port before.
+    ///
+    /// # Panics
+    ///
+    /// This function panics if the provided port is outside the range 1-21.
```

Comment:
Would returning `None` here rather than panicking be more ergonomic?

Reply:
I considered having a check for if the port is out of range, but I decided not to because I didn't see much of a reason. If you think it would be better, then I would be happy to change it.

Reply:
I don't have an opinion either way really. I think it's fine to leave this as-is if nobody else has complaints.